tools/wasi/wasi.cpp

```diff
@@ -35,49 +35,63 @@ ExecutionResult proc_exit(
 
-    const auto fd = args[0].as<uint32_t>();
-    const auto iov_ptr = args[1].as<uint32_t>();
-    const auto iov_cnt = args[2].as<uint32_t>();
-    const auto nwritten_ptr = args[3].as<uint32_t>();
-
-    std::vector<uvwasi_ciovec_t> iovs(iov_cnt);
-    // TODO: not sure what to pass as end, passing memory size...
-    uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(
-        instance.memory->data(), instance.memory->size(), iov_ptr, iovs.data(), iov_cnt);
-    if (ret != UVWASI_ESUCCESS)
-        return Value{uint32_t{ret}};
+        const auto fd = args[0].as<uint32_t>();
+        const auto iov_ptr = args[1].as<uint32_t>();
+        const auto iov_cnt = args[2].as<uint32_t>();
+        const auto nwritten_ptr = args[3].as<uint32_t>();
 
-    uvwasi_size_t nwritten;
-    ret = uvwasi->fd_write(static_cast<uvwasi_fd_t>(fd), iovs.data(), iov_cnt, &nwritten);
-    uvwasi_serdes_write_uint32_t(instance.memory->data(), nwritten_ptr, nwritten);
+        std::vector<uvwasi_ciovec_t> iovs(iov_cnt);
+        // TODO: not sure what to pass as end, passing memory size...
+        uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(
+            instance.memory->data(), instance.memory->size(), iov_ptr, iovs.data(), iov_cnt);
+        if (ret != UVWASI_ESUCCESS)
+            return Value{uint32_t{ret}};
 
-    return Value{uint32_t{ret}};
+        uvwasi_size_t nwritten;
+        ret = uvwasi->fd_write(static_cast<uvwasi_fd_t>(fd), iovs.data(), iov_cnt, &nwritten);
+        uvwasi_serdes_write_uint32_t(instance.memory->data(), nwritten_ptr, nwritten);
+
+        return Value{uint32_t{ret}};
+    }
+    catch (...)
+    {
+        return Value{uint32_t{UVWASI_ENOSYS}};
+    }
 }
 
 ExecutionResult fd_read(
     std::any& host_context, Instance& instance, const Value* args, ExecutionContext&) noexcept
 {
-    auto* uvwasi = *std::any_cast<UVWASI*>(&host_context);
+    try
+    {
+        auto* uvwasi = *std::any_cast<UVWASI*>(&host_context);
 
-    const auto fd = args[0].as<uint32_t>();
-    const auto iov_ptr = args[1].as<uint32_t>();
-    const auto iov_cnt = args[2].as<uint32_t>();
-    const auto nread_ptr = args[3].as<uint32_t>();
-
-    std::vector<uvwasi_iovec_t> iovs(iov_cnt);
-    // TODO: not sure what to pass as end, passing memory size...
-    uvwasi_errno_t ret = uvwasi_serdes_readv_iovec_t(
-        instance.memory->data(), instance.memory->size(), iov_ptr, iovs.data(), iov_cnt);
-    if (ret != UVWASI_ESUCCESS)
-        return Value{uint32_t{ret}};
+        const auto fd = args[0].as<uint32_t>();
+        const auto iov_ptr = args[1].as<uint32_t>();
+        const auto iov_cnt = args[2].as<uint32_t>();
+        const auto nread_ptr = args[3].as<uint32_t>();
 
-    uvwasi_size_t nread;
-    ret = uvwasi->fd_read(static_cast<uvwasi_fd_t>(fd), iovs.data(), iov_cnt, &nread);
-    uvwasi_serdes_write_uint32_t(instance.memory->data(), nread_ptr, nread);
+        std::vector<uvwasi_iovec_t> iovs(iov_cnt);
```

Comment:
Clang-tidy was complaining about exceptions, it turns out both [`std::any_cast`](https://en.cppreference.com/w/cpp/utility/any/any_cast) (in some overloads) and the `std::vector` constructor can yield exceptions.

Reply:
`any_cast` that we use here is `noexcept`, but `vector` does throw, yes.